Performance: Add support for configuring date parser languages

Merinorus opened the pull request:
<!--
Note: All PRs with code changes should be targeted to the `dev` branch, pure documentation changes can target `main`
-->

## Proposed change

Introduce the "PAPERLESS_DATE_PARSER_LANGUAGES" setting, which allows specifying a list of languages supported by the dateparser Python package. Setting the language to avoid automatic detection improves the date parser performance significantly: depending on the date format, detection could take several seconds and is now almost immediate.

This is one of the changes mentioned in this discussion: https://github.com/paperless-ngx/paperless-ngx/discussions/10156 . I prefer making multiple small PRs rather than a big one, as it's easier to review.

<!--
Please include a summary of the change and which issue is fixed (if any) and any relevant motivation / context. List any dependencies that are required for this change. If appropriate, please include an explanation of how your proposed change can be tested. Screenshots and / or videos can also be helpful if appropriate.
-->

<!--
⚠️ Important: Pull requests that implement a new feature or enhancement *should almost always target an existing feature request* with evidence of community interest and discussion. This is in order to balance the work of implementing and maintaining new features / enhancements. If that is not currently the case, please open a feature request instead of this PR to gather feedback from both users and the project maintainers.
-->

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

<!--
NOTE: PRs that do not address the following will not be merged, please do not skip any relevant items.
-->

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

Merinorus committed:
Add support for configuring date parser languages

Introduce the "PAPERLESS_DATE_PARSER_LANGUAGES" setting, which allows
specifying a list of languages supported by the dateparser python package.
Setting the language avoid automatic detection and improves
significantly the date parser performance.

shamoon commented on src/paperless/settings.py:
```diff
@@ -1115,6 +1116,11 @@ def default_threads_per_worker(task_workers) -> int:
 # Specify the default date order (for autodetected dates)
 DATE_ORDER = os.getenv("PAPERLESS_DATE_ORDER", "DMY")
 FILENAME_DATE_ORDER = os.getenv("PAPERLESS_FILENAME_DATE_ORDER")
+DATE_PARSER_LANGUAGES = list(
+    LocaleDataLoader().get_locale_map(
+        languages=json.loads(os.getenv("PAPERLESS_DATE_PARSER_LANGUAGES", "[]")),
+    ),
+)
```
This does seem like an easy win with not a lot of trade-off. That said, making it 100% opt-in seems like a missed opportunity. I think it's worth making this default to the OCR languages while still allowing for manual control. I understand they are not the same, but I suspect we can interpolate between them without too much misery (famous last words?)

stumpylog replied:
Pycountry is a pretty heavy dependent for just one thing.  I think the idea was more a manual interpolation, similar to how nltk is configured with some conversions between formats

shamoon replied:
Yea I’d agree. Don’t want to make you chase your tail but was thinking given how small the sets are here a kind of manual conversion would be feasible. Of course I haven’t done that so perhaps I’m wrong. But I don’t think it’s “worth” a new dep. let us know if you’re up for the challenge (or not)!

Merinorus committed:
Infer date parser languages from OCR_LANGUAGE by default

If "PAPERLESS_DATE_PARSER_LANGUAGES" is not explicitly set, its value
will now be inferred from PAPERLESS_OCR_LANGUAGE.

Format is also changed from a list to a "+" separated string,
to be consistent with the PAPERLESS_OCR_LANGUAGE setting.
Example: '["fr", "de"]' -> "fr+de".

Merinorus committed:
Date parsing: replace pycountry with manual language mapping

Replaced the pycountry dependency with a manual dictionary to map OCR language codes to dateparser-compatible languages. This reduces external dependencies.

Merinorus committed:
Date parser languages: apply code review suggestions

- If an OCR language element is not supported by dateparser,
  fall back to other supported languages instead of disabling language support ([]).
- If a Chinese locale is specified, also provide "zh" to dateparser languages
- Update log messages accordingly.
- Move the Tesseract-to-dateparser language map to a separate module.
- Fix and clarify tests and comments.

Co-authored-by: shamoon <4887959+shamoon@users.noreply.github.com>

Merinorus commented on src/paperless/settings.py:
```diff
@@ -1117,6 +1126,170 @@ def default_threads_per_worker(task_workers) -> int:
 FILENAME_DATE_ORDER = os.getenv("PAPERLESS_FILENAME_DATE_ORDER")
 
+
+def _ocr_to_dateparser_languages(ocr_languages: str) -> list[str]:
```
Noted. I've encapsulated it inside a function because it's used only once, so it can be removed by the garbage collector later.

shamoon commented on src/paperless/tests/test_settings.py:
```diff
@@ -471,3 +473,45 @@ def test_subpath_with_explicit_logout_url(self):
+            _parse_dateparser_languages("en-001+fr-CA"),
+            ["en-001", "fr-CA"],
+        )
+        self.assertEqual(_parse_dateparser_languages("zh-Hans-HK"), ["zh-Hans-HK"])
```
Ha, funny we caught it like this, thank you for tracing this down.

Totally agree with your conclusion seems an issue in dateparser but I think your proposal is perfect:

- Append "zh" when either "zh-Hans" or "zh-Hant" is configured
- Maybe add a warning log to explain the fallback
✅ Document it in configuration.md with a link to the upstream issue

If you (or someone else fluent in Chinese) ends up upstreaming the fix to dateparser, great, but given the age of that issue I'm not holding my breath 😅

shamoon commented on docs/configuration.md:
```diff
@@ -968,6 +968,26 @@ still perform some basic text pre-processing before matching.
+
+!!! note
+This format differs from the `PAPERLESS_OCR_LANGUAGE` setting, which uses ISO 639-2 codes (3 letters, e.g., "eng+deu" for Tesseract OCR).
+
+!!! note
+There is a known [issue](https://github.com/scrapinghub/dateparser/issues/875) with `dateparser` related to Chinese locales: many dates are not correctly detected.
+To work around this, if you specify a Chinese locale such as `"zh-Hant"` or `"zh-Hans-HK"`, the generic Chinese language code `"zh"` will also be added automatically.
```
```suggestion
```

I totally understand the thought process of including this but tbh I think it's not worth it. If anyone is somehow tripped up by this (I doubt) we can always add a note later. Better to keep the docs short overall though imho.

shamoon commented on src/paperless/settings.py:
```diff
@@ -1118,6 +1129,84 @@ def default_threads_per_worker(task_workers) -> int:
+def _parse_dateparser_languages(languages: str | None):
+    language_list = languages.split("+") if languages else []
+    # There is an unfixed issue in zh-Hant and zh-Hans locales.
+    # See: https://github.com/scrapinghub/dateparser/issues/875"""
```
```suggestion
    # See: https://github.com/scrapinghub/dateparser/issues/875
```

shamoon reviewed (approved):
Other than these two tiny things this LGTM, I'll hold off on merging in case our other dev has time / has any other thoughts.

Thanks
